module/scheduler_components/remaster_manager.cpp

```diff
@@ -0,0 +1,159 @@
+    shared_ptr<Storage<Key, Record>> storage,
+    shared_ptr<TransactionMap> all_txns)
+  : config_(config), storage_(storage), all_txns_(all_txns) {}
+
+VerifyMasterResult
+RemasterManager::VerifyMaster(const TxnReplicaId txn_replica_id) {
+  auto& txn_holder = all_txns_->at(txn_replica_id);
+  auto txn = txn_holder.GetTransaction();
+  auto& keys = txn_holder.KeysInPartition();
+  if (keys.empty()) {
+    // None of the keys in this txn are in this partition
+    return VerifyMasterResult::VALID;
+  }
+
+  auto& txn_master_metadata = txn->internal().master_metadata();
+  // This should only be the case for testing
+  if (txn_master_metadata.empty()) {
+    LOG(WARNING) << "Master metadata empty: txn id " << txn->internal().id();
+    return VerifyMasterResult::VALID;
+  }
+
+  // keeps track of counters that will need to be waited on
+  std::unordered_map<Key, int32_t> ahead_counters;
+  for (auto& pair : keys) {
+    auto key = pair.first;
+
+    // Check with storage to get current counter
+    if (counters_.count(key) == 0) {
+      Record record;
+      bool found = storage_->Read(key, record);
+      if (found) {
+        CHECK(txn_master_metadata.contains(key))
+              << "Master metadata for key \"" << key << "\" is missing";
+        counters_[key] = record.metadata.counter;
+      } else {
+        counters_[key] = 0;
+      }
+    }
+    auto stored_counter = counters_.at(key);
+    auto txn_metadata = txn_master_metadata.at(key);
+    if (txn_metadata.counter() > stored_counter) {
+      ahead_counters[key] = txn_metadata.counter();
+    } else if (txn_metadata.counter() < stored_counter) {
+      return VerifyMasterResult::ABORT;
+    }
+  }
+
+  // Check if no keys are blocked at all
+  if (ahead_counters.empty()){
+    auto no_key_indirectly_blocked = true;
+    for (auto& pair : keys) {
+      auto key = pair.first;
+      // If the queue is emtpy, can be skipped
+      if (blocked_queue_.count(key)) {
+        // If we won't be at the front of the queue, the txn will need to wait.
+        if (blocked_queue_.at(key).front().second <= txn_master_metadata.at(key).counter()) {
+          no_key_indirectly_blocked = false;
+          break;
+        }
+      }
+    }
+    if (no_key_indirectly_blocked) {
+      return VerifyMasterResult::VALID;
+    }
+  }
+
+  // add the txn to the queue for each key
+  for (auto& pair : keys) {
+    auto key = pair.first;
+    auto counter = txn_master_metadata.at(key).counter();
+    InsertIntoBlockedQueue(key, counter, txn_replica_id);
+  }
+  return VerifyMasterResult::WAITING;
+}
+
+void RemasterManager::InsertIntoBlockedQueue(const Key key, const uint32_t counter, const TxnReplicaId txn_replica_id) {
+  auto entry = make_pair(txn_replica_id, counter);
+
+  // Iterate until at end or counter is smaller than next element. Maintains priority queue
+  auto itr = blocked_queue_[key].begin();
+  while (itr != blocked_queue_[key].end() && (*itr).second <= counter) {
```

Comment:
nit: The `=` of `<=` here is very important to maintain the order of the txn having the same version in a local log, so I think there should be comment about it here.